### PR TITLE
[Speed]Refine ParallelExecutor

### PR DESCRIPTION
**Pull request by chengduoZH**
related issue: https://github.com/PaddlePaddle/Paddle/issues/15049

The original implementation of `OpHandle::Run` is inefficient, i.e. the ncclAllReduce of the 0th card must wait for the inputs of all cards are ready. In fact, it is unnecessary, the 0th card only need wait for the input of the 0th card is ready.

The work of this PR includes:
1. Add Event in VarHandle.
2. Record the corresponding Event to VarHandle.
3. Refine `OpHandleBase::WaitInputVarGenerated`, before OpHandle runs, make the current stream only wait the input VarHandle's Event(**but not the DummyVarHandle's Event**) , i.e. `cudaStreamWaitEvent(stream, in_var_handle->GetEvent(), 0))`.  Because DummyVarHandle is only used to record the dependence of OpHandles.

The below table displays the current speed of the `parallel graph executor` and `experimental executor` on Transformer model and ResNet model.

  | ResNet | ResNet | Transformer | Transformer
-- | -- | -- | -- | --
  | 4 cards (s/batch) | 8cards (s/batch) | 4cards(step/s) | 8cards (step/s)
parallel graph executor | 0.1017 | 0.1462 | 3.696671 | 2.67
experimental executor | 0.1066 | 0.1658 | 3.65736 | 2.613756
before | 0.149 | 0.1935 | 3.276035 | 2.255209

As the table says that this PR makes the experimental executor faster than before, but also a bit slower than `parallel graph executor`, I'll do some in-depth performance analysis next.

To reproduce the above performance data, you need to add dependence of all the all_reduce_op_handle and add sync on nccl stream in all_reduce_op_handle. Because `all_reduce_deps_pass` has bugs in other models, I add an environment variable to use `all_reduce_deps_pass` or not, i.e `env GLOG_vmodule=build_strategy=2`. After the bugs of `all_reduce_deps_pass` are fixed, I will remove this environment variable.

### CODE REVIEW

File: paddle/fluid/framework/details/all_reduce_op_handle.cc [hunk 1]
```diff
@@ -55,20 +55,27 @@ AllReduceOpHandle::AllReduceOpHandle(ir::Node *node,
...
+      place_vars[in_var->place()].emplace_back(in_var);
+    }
+  }
+
   std::vector<const LoDTensor *> lod_tensors;
   for (size_t i = 0; i < local_scopes_.size(); ++i) {
-    auto *s = local_scopes_[i];
-    auto &local_scope = *s->FindVar(kLocalExecScopeName)->Get<Scope *>();
+    Scope &local_scope =
+        *local_scopes_[i]->FindVar(kLocalExecScopeName)->Get<Scope *>();
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/all_reduce_op_handle.cc]: FindVar may still do `hash(kLocalExecScopeName)`, use static list should save small amount of time.

File: paddle/fluid/framework/details/op_handle_base.cc [hunk 2]
```diff
@@ -78,6 +110,39 @@ void OpHandleBase::RecordWaitEventOnCtx(platform::DeviceContext *waited_ctx) {
 #endif
 }
 
+void OpHandleBase::RecordWaitEventOnCtx2(
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/op_handle_base.cc]: Why need to add this? Note that experiments show that in distributed training with nccl2, wait allreduce to finish slows down the total throughput.

File: paddle/fluid/framework/details/all_reduce_op_handle.cc (hunk 1 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/all_reduce_op_handle.cc]: `LocalExecScope` can be dropped in `ScopeBufferedSSAGraphExecutor`, so it is inappropriate that save them in a static list.

File: paddle/fluid/framework/details/op_handle_base.cc (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/op_handle_base.cc]: `RecordWaitEventOnCtx2` is just a test function, the origin implementation of all_reduce_op_handle is not efficiently, and this PR is fixing it.

File: paddle/fluid/framework/details/all_reduce_op_handle.cc [hunk 3]
```diff
@@ -97,9 +104,12 @@ void AllReduceOpHandle::RunImpl() {
 
       int dev_id = boost::get<platform::CUDAPlace>(p).device;
       auto &nccl_ctx = nccl_ctxs_->at(dev_id);
+      auto dev_ctx = nccl_ctxs_->DevCtx(p);
+      auto &vars = place_vars[p];
       auto stream = nccl_ctx.stream();
       auto comm = nccl_ctx.comm_;
       all_reduce_calls.emplace_back([=] {
+        RecordWaitEventOnCtx2(vars, dev_ctx);
```
Comment (panyx0718) [inline paddle/fluid/framework/details/all_reduce_op_handle.cc]: why?

File: paddle/fluid/framework/details/build_strategy.cc [hunk 4]
```diff
@@ -122,6 +122,10 @@ class ParallelExecutorPassBuilder : public ir::PassBuilder {
     // Verify that the graph is correct for multi-device executor.
     AppendPass("multi_devices_check_pass");
 
+    if (VLOG_IS_ON(2)) {
```
Comment (panyx0718) [inline paddle/fluid/framework/details/build_strategy.cc:150]: ?

File: paddle/fluid/framework/details/fetch_op_handle.cc [hunk 5]
```diff
@@ -44,7 +44,10 @@ void FetchOpHandle::WaitAndMergeCPUTensors() const {
 }
 
 void FetchOpHandle::RunImpl() {
-  WaitInputVarGenerated(platform::CPUPlace());
+  auto cpu_ctx =
+      platform::DeviceContextPool::Instance().Get(platform::CPUPlace());
+  auto in_var_handles = DynamicCast<VarHandle>(this->Inputs());
+  RecordWaitEventOnCtx2(in_var_handles, cpu_ctx);
```
Comment (panyx0718) [inline paddle/fluid/framework/details/fetch_op_handle.cc]: why?

File: paddle/fluid/framework/details/op_handle_base.h [hunk 6]
```diff
@@ -46,30 +48,22 @@ class OpHandleBase {
 
   virtual void RecordWaitEventOnCtx(platform::DeviceContext *waited_ctx);
 
+  virtual void RecordWaitEventOnCtx2(const std::vector<VarHandle *> &in_vars,
```
Comment (panyx0718) [inline paddle/fluid/framework/details/op_handle_base.h]: any doc?

**Review by panyx0718 — COMMENTED** [review]
why change, what changed, what improved?

File: paddle/fluid/framework/details/all_reduce_op_handle.cc (hunk 3 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/all_reduce_op_handle.cc]: the original implementation is inefficient, i.e. the `ncclAllReduce` of the 0th card must wait for the inputs of all cards are ready. in fact, it is unnecessary, the 0th card only need wait for the input of the 0th card is ready.

File: paddle/fluid/framework/details/build_strategy.cc (hunk 4 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/build_strategy.cc:150]: It is just a test mode.

Comment (Yancey0623) [reply paddle/fluid/framework/details/build_strategy.cc:150]: confusion about the VLOG... using the GFLAGS?

if (SeqOnlyAllReduceOps(strategy) || FLAGS_seq_allreduce_pass) ?

**Review by Yancey0623 — COMMENTED** [review]
Maybe we need an experiment on a model such as the ResNet, and show the speed improment.

**Review by Yancey0623 — APPROVED** [review]
LGTM!

**Pull request merged**